src/topology_sampler.hpp

```diff
@@ -0,0 +1,129 @@
+// Copyright 2019-2021 bito project contributors.
+// bito is free software under the GPLv3; see LICENSE file for details.
+
+#pragma once
+
+#include "node.hpp"
+#include "mersenne_twister.hpp"
+#include "eigen_sugar.hpp"
+
+class TopologySampler {
+public:
+
+  class Input {
+  public:
+    virtual EigenConstVectorXdRef SBNParameters() const = 0;
+    virtual size_t RootsplitCount() const = 0;
+    virtual const Bitset &RootsplitsAt(size_t rootsplit_idx) const = 0;
+    virtual const SizePair &ParentToRangeAt(const Bitset &parent) const = 0;
+    virtual const Bitset &IndexToChildAt(size_t child_idx) const = 0;
+  };
+
+  Node::NodePtr SampleTopology(const Input& input, bool is_rooted) const;
+  inline void SetSeed(uint64_t seed) { mersenne_twister_.SetSeed(seed); }
+
+private:
+  using Range = std::pair<size_t, size_t>;
+
+  size_t SampleIndex(const Input &input, Range range) const;
+  Node::NodePtr SampleTopology(const Input &input, const Bitset &parent_subsplit) const;
+  
+  MersenneTwister mersenne_twister_;
+};
+
+template <typename T>
+class SBNInstanceSamplerInput : public TopologySampler::Input {
+public:
+  SBNInstanceSamplerInput(const T &inst) :
+    inst_{inst} {}
+
+  size_t RootsplitCount() const override {
+    return inst_.SBNSupport().RootsplitCount();
+  }
+
+  EigenConstVectorXdRef SBNParameters() const override {
+    return inst_.SBNParameters();
+  }
+  
+  const Bitset &RootsplitsAt(size_t rootsplit_idx) const override {
+    return inst_.SBNSupport().RootsplitsAt(rootsplit_idx);
+  }
+  
+  const SizePair &ParentToRangeAt(const Bitset &parent) const override {
+    return inst_.SBNSupport().ParentToRangeAt(parent);
+  }
+  
+  const Bitset &IndexToChildAt(size_t child_idx) const override {
+    return inst_.SBNSupport().IndexToChildAt(child_idx);
+  }
+  
+private:
+  const T &inst_;
+};
+
+class SubsplitDAG;
+class SubsplitDAGSamplerInput : public TopologySampler::Input {
+public:
+  SubsplitDAGSamplerInput(const SubsplitDAG& dag, EigenConstVectorXdRef sbn_parameters) :
+    dag_{dag}, sbn_parameters_{sbn_parameters} {}
+
+  size_t RootsplitCount() const override;
+  EigenConstVectorXdRef SBNParameters() const override;
+  const Bitset &RootsplitsAt(size_t rootsplit_idx) const override;
+  const SizePair &ParentToRangeAt(const Bitset &parent) const override;
+  const Bitset &IndexToChildAt(size_t child_idx) const override;
+
+private:
+  const SubsplitDAG &dag_;
+  EigenConstVectorXdRef sbn_parameters_;
+};
+
+#ifdef DOCTEST_LIBRARY_INCLUDED
+
+#include "doctest_constants.hpp"
+#include "unrooted_sbn_instance.hpp"
```

Comment:
Wow!!

If I understand correctly, this doesn't set up a circular dependency because this code isn't part of `topology_sampler.hpp` and really just gets pulled into the doctest. Is that right?

Reply:
There isn't a circular dependency, but it's written like this to avoid bringing in unnecessary symbols during normal usage (outside doctest). If there was a circular dependency, it would be triggered during doctest compilation.